### ISSUE TITLE
Memoized queries are not cleared by changes to properties

### ISSUE THREAD

**Issue by Zirro** [tmpvar/jsdom#1835]
### Basic info:

- **Node.js version: 7.7.3** 
- **jsdom version: 10.1.0** 

### Minimal reproduction case

```js
const { JSDOM } = require("jsdom");

const dom = new JSDOM(`
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <title>HelloWorld</title>
</head>
<body>
<select multiple id="foo">
  <option>one</option>
  <option>two</option>
  <option selected>three</option>
</select>
</body>
</html>
`);

const el = dom.window.document.querySelector("#foo");

const test1 = el.querySelector(":checked");

el.lastElementChild.selected = false;

const test2 = el.querySelector(":checked");

console.log(test1, test2);

```

### How does similar code behave in browsers?

https://jsfiddle.net/yyur6236/

In browsers, `test1` will be an `<option>` and `test2` will be `null`.

In jsdom, `test1` will be an `<option>` and test2 will be equal to test1 - the same `<option>` which is no longer selected.

I've tracked this down to the memoizeQuery function which is used on querySelector() and similar methods. `Node-impl.js` has a `_clearMemoizedQueries()` method which runs as part of `_modified()`.

I figured that I could potentially resolve this by calling `super._modified()` in any setter which could have an effect on memoized queries, though I could only find [one instance](https://github.com/tmpvar/jsdom/blob/master/lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js#L32) of this call being made before.

My question is whether I ought to start adding `super._modified()` to certain setters, or if there is an alternative solution which is preferable?

**Comment by domenic** [comment]
Great catch!

I think you're right that any time we change state that could affect CSS matching, we need to call `this._modified()` (not `super`, I am pretty sure). Auditing for those will be a bit tricky, as it's not just the public setters; e.g. form reset and more can affect things.

**Comment by Zirro** [comment]
You're right, `this._modified()` will do. I'll start by going through the state-based selectors one-by-one to determine the properties which they depend on, and locate any functions changing those properties in each impl.js-file.
